### PR TITLE
♻️⚡️ Refactor FlowNodeHandling to Promise-Chains

### LINKED ISSUE #212: 🐛 ParallelGateways are unstable
#### Describe your issue

There are currently multiple, quite severe issues with the ParallelGateway-Handler:
- There is no support for nested Gateways, either for additional ParallelGateways, or ExclusiveGateways. Use any and the whole thing breaks.
- It is always assumed that every branch from the Split-Gateway will ultimately lead back to a Join-Gateway. This is obviously not the case for EndEvents.
- The ProcessModelFacade, which performs the search for a corresponding Join-Gateway, cannot track branches reliably through nested Gateways,
because it always assumes that each FlowNode in a branch has only ever one outgoing SequenceFlow.
This is obviously not the case with Gateways  (which have multiple) or EndEvents (which have none).
- All branch execution is handled within the Split-Gateway. The Join-Gateway doesn't do anything. This effectively moves the execution of FlowNodes to the SplitGateway and away from the Execution Service, meaning we now have two points of execution to take care of.

We urgently need to come up with a better way for handling these Gateways, because in their current state, they are pretty much unusable.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### LINKED ISSUE #225: ♻️⚡️ Refactor handling of FlowNodeInstances to Promise-Chain based handling
#### Describe your issue

Currently the responsibility for executing FlowNodeInstances resides exclusively within the `ExecuteProcessService`. 
The FlowNodeHandlers only execute their own stuff and then return with a so called `NextFlowNodeInfo`, which the `ExecuteProcessService` can then use to run the next FlowNodeInstance.

While this works well enough, it also effectively prevents us from implementing a few features, such as cyclic `IntermediateTimerEvents`/`TimerStartEvents` or nested `ParallelGateways`.

This is because the `ExecuteProcessService` runs each FlowNodeInstance only once and then moves on to the next.
This makes it impossible to use cyclic timers, because while the timer instance may still be running, the FlowNode that follows after it, will only ever be executed once.

For the ParallelGateways, this has the effect of having to run every parallel branch within the `ParallelSplitGateway` handler, which rather breaks the pattern and leads to a duplication of the `ExecuteProcessService` logic.

#### Possible solution

Instead of using a centralized point of execution, we should switch to a Promise-Chain-based logic.
The `ExecuteProcessService` should only start the execution of the `ProcessInstances` StartEvent and then wait for a notification about the `ProcessInstances` end.

In essence, the Promise Chain would look like this:

```ts
class ExecuteProcessService {

  public async execute(): Promise<someType> {
    return StartEvent.execute(); 
  }
}

class StartEvent {

  public async execute(): Promise<someType> {
    return SomeFlowNode.execute();
  }
}

class SomeFlowNode {

  public async execute(): Promise<someType> {
    return EndEvent.execute();
  }
}

class EndEvent {

  public async execute(): Promise<someType> {
    eventAggregator.publish('ProcessInstanceFinished', someStuff);
    return Promise.resolve(someStuff);
  }
}
```

This is of course just a very simplistic demonstration of what will actually be happening, but it should convey the general idea.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by ElRaptorus**
**Requires:**
https://github.com/process-engine/process_engine_contracts/pull/100

**Changes:**

1. Remove all `execute` and `resume` logic from the `ExecuteProcessService` and `ResumeProcessService`.
    - The services no longer handle any kind of FlowNode execution.
    - The handlers now do this in a Promise chain, as described [here](https://github.com/process-engine/process_engine_core/issues/225).
2. Centralize the Subscription for `TerminateEndEvent` notifications to the `FlowNodeHandlerInterruptible` base handler.
    - This has the effect of providing all FlowNodeHandlers to subscribe for such a notification and react appropriately when it arrives.
3. Add `beforeExecute` and `afterExecute` hooks to the base FlowNodeHandler, which will allow derived handlers to perform customized setup- and cleanup- operations.
    - Handling subscriptions for TerminateEndEvents would be one application for these.
4. Remove the parent-handler pattern for `ServiceTasks`, `ParallelGateways`, `IntermediateCatchEvents` and `IntermediateThrowEvents`.
    - Dedicated FlowNodeHandlerFactories will now be used to create the various handler for those. 
5. Refactor all FlowNodeHandlers to return a list of `NextFlowNodes`, instead of just a single one. 
    - For most FlowNode types, only one outgoing SequenceFlow is still permitted.
    - Any type of Gateway can return as many SequenceFlows as they want (with the Exception of the ExclusiveGateway, which by design only returns one).
    - The ProcessModelFacade has been refactored to ensure that only Gateways will be allowed to have multiple `NextFlowNodes`.
    - This change makes it very easy to run parallel branches for a ParallelGateway.
6. Refactor ParallelGateways:
    - `ParallelSplitGateways` now do nothing but return a list of the FlowNodes starting each branch.
    - For each `ParallelJoinGateway`, a single instance is created and stored at the IoC Container.
    - The number of times such an instance can be called is equal to the incoming SequenceFlows attached to it.
    - Each time the instance is called, the result from the preceeding FlowNodeInstance is stored.
    - As soon as the expected number of results was received, the instance will resume the ProcessInstance with the next FlowNode and remove itself from the IoC container.
7. Implement `FlowNodePersistenceFacade`, to decouple the FlowNodeHandlers from the persistence logic.
8. Refactor BoundaryEvents to use a base `BoundaryEventHandler` class and decouple the handlers from their decorated handler.
    - In the old system, the BoundaryEvents executed the handler they were attached to. With the new chaining system, this chain has been reversed.
    - The BoundaryEvents get initialized by the handler they are attached to and trigger a callback each time their triggering event occurs.
    - The decorated handler also discards them, after it has finished execution.
    - This finally allows us to use non-interrupting BoundaryEvents and cyclic timers, because the decorated handlers can effectively run independently from their BoundaryEvents, but can also be interrupted by them, if need be.
9. Move creation of BoundaryEventHandlers into a dedicated factory.
10. Use customized error-type to handle errors thrown by an ErrorEndEvent (Bluebird doesn't event allow to reject non-errors).
11. Fix `cancelActivity` property parsing for BoundaryEvents
12. Cleanup `ProcessTokenFacade`:
    - Remove obsolete mapper functions
    - Expect a FlowNodeInstanceId to be passed to `addResultForFlowNode`
    - Add `containsResultForFlowNodeInstance` checker function

**Issues:**

Closes #212 
Closes #225
Part of #213
Part of https://github.com/process-engine/process_engine_runtime/issues/155
Part of https://github.com/process-engine/process_engine_runtime/issues/163
Part of https://github.com/process-engine/process_engine_runtime/issues/187
Part of https://github.com/process-engine/process_engine_runtime/issues/197

PR: #239

## How can others test the changes?

All in all, everything should run as before.
The only notable change should be in the ParallelGatewayHandlers, which should be a lot more reliable now.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

### CODE REVIEW

File: ioc_module.js [hunk 1]
```diff
@@ -95,157 +97,180 @@ function registerServices(container) {
...
+      'container',
+      'CorrelationService',
+      'ResumeProcessService',
+    );
 
   container
-    .register('IntermediateCatchEventHandler', IntermediateCatchEventHandler)
-    .dependencies('container', 'FlowNodeInstanceService', 'LoggingApiService', 'MetricsApiService');
+    .register('EndEventHandler', EndEventHandler)
+    .dependencies('container');
```
Comment (roblen45) [inline ioc_module.js]: This is not the solution for your problem. 
Your problem are too many dependencies. 
Hiding dependencies is not the solution.

See:

http://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/

I would prefer to encapsulate the technical dependencies (logging, persistence, metrics) in one dependencies.

Same problem with every FlowNodeHandler.

File: src/runtime/engine/flow_node_handler/intermediate_event_handlers/intermediate_catch_event_factory.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import {IContainer} from 'addict-ioc';
+
+import {UnprocessableEntityError} from '@essential-projects/errors_ts';
+import {
+  IFlowNodeHandler,
+  IFlowNodeHandlerDedicatedFactory,
+  Model,
+} from '@process-engine/process_engine_contracts';
+
+import {FlowNodeHandlerInterruptible} from '../flow_node_handler_interruptible';
```
Comment (roblen45) [inline src/runtime/engine/flow_node_handler/intermediate_event_handlers/intermediate_catch_event_factory.ts]: We have to talk about FlowNodeHandlerInterruptible.

I think the boundary events should not be part of this class.

File: src/runtime/engine/flow_node_handler/intermediate_event_handlers/intermediate_link_catch_event_handler.ts [hunk 3]
```diff
@@ -1,53 +1,50 @@
+import {IContainer} from 'addict-ioc';
```
Comment (roblen45) [inline src/runtime/engine/flow_node_handler/intermediate_event_handlers/intermediate_link_catch_event_handler.ts]: Dependency of IoC in business logic!!

File: ioc_module.js (hunk 1 above)
Comment (ElRaptorus) [reply ioc_module.js]: I did that with a `FlowNodePersistenceFacade`. 

I'll readjust the registrations again.

File: src/runtime/engine/flow_node_handler/intermediate_event_handlers/intermediate_catch_event_factory.ts (hunk 2 above)
Comment (ElRaptorus) [reply src/runtime/engine/flow_node_handler/intermediate_event_handlers/intermediate_catch_event_factory.ts]: Yeah, I reflected on this during the weekend. You're correct.

**Pull request merged**